[git-webkit] add option to update list of changed files

briannafan opened the pull request:
#### d68e12aac8d992e36259542a6ed9677b0a60cef3
<pre>
[git-webkit] add option to update list of changed files
<a href="https://bugs.webkit.org/show_bug.cgi?id=285368">https://bugs.webkit.org/show_bug.cgi?id=285368</a>
<a href="https://rdar.apple.com/140809288">rdar://140809288</a>

Reviewed by Jonathan Bedard.

Add a feature that automatically updates the changelog in your
commit message with the most recent changes.

This feature will be enabled by default. It can be disabled by:
    1) Set webkitscmpy.auto-update-changelog in .git/config to false
    2) Run git-webkit commit --amend --no-update

* Tools/Scripts/hooks/prepare-commit-msg:
    - When amending a commit with changes, parse the message and update
      the changelog with the modified lines. Then, reconstruct the message.
    - Add the relevant instructions/logging when the changelog is or is not
      updated.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/commit_parser.py:
    - Change regex const names to match the expression (removed &lt;-&gt; deleted)
(CommitMessageParser._parse_modified_files_lines):
(CommitMessageParser.apply_comments_to_modified_files_lines):
    - Add return_deleted arg which will only return the lines that end with &quot;Deleted&quot; or &quot;Removed&quot;.
    - Catch the case in which a comment is made after &quot;Filename: [Added/Removed/Deleted].&quot;
(CommitMessageParser.delete_trailing_blank_lines):
    - Fix class method arguments

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/commit.py:
(Commit.parser): Add --update/--no-update flags
(Commit.main): Set WKSCMPY_UPDATE_CHANGELOG env variable

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
(Setup.git): Set webkitscmpy.auto-update-changelog.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/commit_parser_unittest.py: Added.
(TestCommitParser):
(TestCommitParser.test_basic_commit):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/setup_unittest.py:

Canonical link: <a href="https://commits.webkit.org/295625@main">https://commits.webkit.org/295625@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/45b9c2058f61bebb30268d65698a0d6f6b43bf0c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/96266 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15880 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5849 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/101333 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46785 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/98311 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/16176 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/24313 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73381 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30611 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/99269 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/12137 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86961 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53718 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/95753 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11889 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4723 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/46112 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/82006 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4821 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/103361 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/23333 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16990 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82420 "Found 1 new test failure: imported/w3c/web-platform-tests/service-workers/service-worker/redirected-response.https.html (failure)") | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/95837 "Passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23584 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82981 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81795 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26424 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3862 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16707 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/17140 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/23296 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22955 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/26435 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24696 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -332,22 +347,52 @@ def main(file_name=None, source=None, sha=None):
 
+    update_changelog = False
+    try:
+        update_changelog = subprocess.check_output(
```
We should just be using [`subprocess.run`](https://docs.python.org/3/library/subprocess.html#subprocess.run) nowadays, now we're requiring recent enough Python.

JonWBedard replied:
Can +1 this. The reason we're using `subprocess.check_output(` is because this _was_ Python 2 compatible not that long ago.

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -332,22 +347,52 @@ def main(file_name=None, source=None, sha=None):
+    if os.environ.get('UPDATE_CHANGELOG', '') == 'True':
+        update_changelog = True
+    elif os.environ.get('UPDATE_CHANGELOG', '') == 'False':
+        update_changelog = False
```
Nit: I do slightly wonder if we should be prefixing these environment variables with `WEBKITSCMPY_` or something, per common convention.

briannafan replied:
Could we do something like `WKSCMPY_` for brevity? I can start with the ones in the files I touch.

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -332,22 +347,52 @@ def main(file_name=None, source=None, sha=None):
+    update_changelog = False
+    try:
+        update_changelog = subprocess.check_output(
+            ['git', 'config', '--get', 'webkitscmpy.auto-update-changelog'],
```
```suggestion
            ['git', 'config', '--get', '--type', 'bool', 'webkitscmpy.auto-update-changelog'],
```

Would probably be better, so `git` does its normal canonicalisation.

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -34,6 +35,7 @@ UNPACK_SECONDARY_RE = re.compile(r' \(({})\)'.format(COMMIT_REF_BASE))
 sys.path.append(SCRIPTS)
 from webkitpy.common.checkout.diff_parser import DiffParser
 from webkitbugspy import radar
+from webkitscmpy.commit_parser import CommitMessageParser
```
This will inevitably break the hook if you're on an older branch.

JonWBedard replied:
Indeed! `DiffParser` isn't so bad because it's ancient.

What is our commit message template behavior in if we fail this import?

briannafan replied:
Based on our discussion, I've implemented a fallback to the original commit message template which doesn't update the change logs. It also only tries to import the parser if we're amending and `update_changelog = True` to minimize impact

gsnedders reviewed (commented):
Posting these now, let's talk more about this on Tuesday, and maybe figure out if we want to move logic out of the hooks themselves. As it is, it's quite hard to have confidence that all the untested code here is actually correct.

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -304,6 +312,51 @@ def cherry_pick(content):
+    if amend_changes and update_changelog:
+        try:
+            from webkitscmpy.commit_parser import CommitMessageParser
+            commit_message_parser = CommitMessageParser()
```
We don't need such a long try block here; we can much more narrowly scope what we're catching exceptions for:

```suggestion
        try:
            from webkitscmpy.commit_parser import CommitMessageParser
        except ImportError as e:
            sys.stderr.write(f'{e}: Could not update changelog automatically. Falling back to the default template.\n')
        else:
            commit_message_parser = CommitMessageParser()
```

briannafan replied:
Good point

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -304,6 +312,51 @@ def cherry_pick(content):
+            commit_message_parser.parse_message(full_message)
+            updated_changelog_lines = commit_message_parser.apply_comments_to_modified_files_lines(combined_changes)
+            reviewed_by_lines = (commit_message_parser.reviewed_by_lines or ['Reviewed by NOBODY (OOPS!).']) + ['']
+            description_lines = (commit_message_parser.description_lines or ['Explanation of why this fixes the bug (OOPS!).']) + ['']
```
Does it make sense to append an empty string to the lists here, versus adding in the `amended_message` (as we do after the title)?

briannafan replied:
Right, I think keeping it consistent is better

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -332,28 +385,49 @@ def main(file_name=None, source=None, sha=None):
+    if os.environ.get('WKSCMPY_UPDATE_CHANGELOG', '') == 'True':
+        update_changelog = True
+    elif os.environ.get('WKSCMPY_UPDATE_CHANGELOG', '') == 'False':
+        update_changelog = False
```
The subprocess invocation should probably be in an `else` block; subprocesses are comparatively slow.

briannafan replied:
Good catch. No point checking the config if there is an override

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -332,28 +385,49 @@ def main(file_name=None, source=None, sha=None):
                 commit_message_file.write('# {}\n'.format(line))
             commit_message_file.write('\n')
-        for line in subprocess.check_output(
+        for line in subprocess.run(
```
This needs `check=True` to match prior semantics.

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -5,12 +5,13 @@ import re
 import sys
 
-TOPLEVEL = subprocess.check_output(['git', 'rev-parse', '--show-toplevel'], encoding='utf-8').strip()
+TOPLEVEL = subprocess.run(['git', 'rev-parse', '--show-toplevel'], encoding='utf-8', capture_output=True).stdout.strip()
```
This needs `check=True` to match prior semantics.

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -35,8 +36,9 @@ sys.path.append(SCRIPTS)
+
 def set_env_variables_from_branch_config():
-    BRANCH = subprocess.check_output(['git', 'rev-parse', '--abbrev-ref', 'HEAD'], encoding='utf-8').strip()
+    BRANCH = subprocess.run(['git', 'rev-parse', '--abbrev-ref', 'HEAD'], encoding='utf-8', capture_output=True).stdout.strip()
```
This needs `check=True` to match prior semantics.

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -48,11 +50,11 @@ def set_env_variables_from_branch_config():
             value = ''
             count = 0
-            for line in reversed(subprocess.check_output(
+            for line in reversed(subprocess.run(
```
This needs `check=True` to match prior semantics.

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -93,41 +95,45 @@ def get_bugs_string():
 
     try:
-        for line in subprocess.check_output(
+        for line in subprocess.run(
```
This needs `check=True` to match prior semantics.

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -163,10 +169,11 @@ def source_branches():
         while outer_line:
             branches = set()
-            for inner_line in subprocess.check_output(
+            for inner_line in subprocess.run(
```
This needs `check=True` to match prior semantics.

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -244,16 +251,17 @@ def cherry_pick(content):
     if production_hash:
         try:
-            subprocess.check_output(
+            subprocess.run(
```
This needs `check=True` to match prior semantics.

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -244,16 +251,17 @@ def cherry_pick(content):
         except subprocess.CalledProcessError:
             remotes = set()
-            for line in subprocess.check_output(
+            for line in subprocess.run(
```
This needs `check=True` to match prior semantics.

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -332,28 +385,49 @@ def main(file_name=None, source=None, sha=None):
 
+    update_changelog = False
+    try:
+        update_changelog = subprocess.run(
```
This needs `check=True` or else it will never raise `CalledProcessError`.

gsnedders commented on Tools/Scripts/hooks/prepare-commit-msg:
```diff
@@ -332,28 +385,49 @@ def main(file_name=None, source=None, sha=None):
+        generated_msg, combined_changes, amend_changes = message(source=source, sha=sha)
         if sha:
-            commit_message_file.write(subprocess.check_output(
+            git_log_msg = subprocess.run(
```
This needs `check=True` to match prior semantics.

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/commit.py:
```diff
@@ -98,6 +103,11 @@ def main(cls, args, repository, command=None, representation=None, **kwargs):
+        if args.update:
+            env['WKSCMPY_UPDATE_CHANGELOG'] = 'True'
+        elif args.update is False:
+            env['WKSCMPY_UPDATE_CHANGELOG'] = 'False'
```
Instead of the `is False`, I think I'd go with `if args.update is not None: env['WKSCMPY_UPDATE_CHANGELOG'] = str(bool(args.update))`.

gsnedders reviewed (commented):
`commit_parser` looks like its changes should be relatively easily testable? Because we should definitely have tests for this, it's the sort of thing we're relatively likely to accidentally break.

gsnedders commented on Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/commit_parser_unittest.py:
```diff
@@ -0,0 +1,96 @@
+* File.py: Added. Comment 2.
+''')
+        self.assertEqual(['* Change.py: Comment on the file.', '* File.py: Added. Comment 2.'], commit_message_parser.modified_files_lines)
+        self.assertEqual(['* Change.py: Comment on the file.', '* File.py: Added. Comment 2.', '* new_file.cpp: Added.'], commit_message_parser.apply_comments_to_modified_files_lines(['* Change.py:', '* File.py: Added.', '* new_file.cpp: Added.']))
```
Very much a nit, but can we wrap these so the lines are less long?

I'd probably lean towards the [Black](https://black.readthedocs.io/en/stable/index.html)-formatted:

```python
        self.assertEqual(
            ["* Change.py: Comment on the file.", "* File.py: Added. Comment 2."],
            commit_message_parser.modified_files_lines,
        )

        self.assertEqual(
            [
                "* Change.py: Comment on the file.",
                "* File.py: Added. Comment 2.",
                "* new_file.cpp: Added.",
            ],
            commit_message_parser.apply_comments_to_modified_files_lines(
                ["* Change.py:", "* File.py: Added.", "* new_file.cpp: Added."]
            ),
        )
```

Or even add trailing commas to the shorter lists to force them to also get wrapped like:

```python
        self.assertEqual(
            [
                "* Change.py: Comment on the file.",
                "* File.py: Added. Comment 2.",
            ],
            commit_message_parser.modified_files_lines,
        )

        self.assertEqual(
            [
                "* Change.py: Comment on the file.",
                "* File.py: Added. Comment 2.",
                "* new_file.cpp: Added.",
            ],
            commit_message_parser.apply_comments_to_modified_files_lines(
                [
                    "* Change.py:",
                    "* File.py: Added.",
                    "* new_file.cpp: Added.",
                ]
            ),
        )
```

To me at least, it's a lot easier to read what's being compared like this.

briannafan committed:
[git-webkit] add option to update list of changed files
https://bugs.webkit.org/show_bug.cgi?id=285368
rdar://140809288

Reviewed by Jonathan Bedard.

Add a feature that automatically updates the changelog in your
commit message with the most recent changes.

This feature will be enabled by default. It can be disabled by:
    1) Set webkitscmpy.auto-update-changelog in .git/config to false
    2) Run git-webkit commit --amend --no-update

* Tools/Scripts/hooks/prepare-commit-msg:
    - When amending a commit with changes, parse the message and update
      the changelog with the modified lines. Then, reconstruct the message.
    - Add the relevant instructions/logging when the changelog is or is not
      updated.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/commit_parser.py:
    - Change regex const names to match the expression (removed <-> deleted)
(CommitMessageParser._parse_modified_files_lines):
(CommitMessageParser.apply_comments_to_modified_files_lines):
    - Add return_deleted arg which will only return the lines that end with "Deleted" or "Removed".
    - Catch the case in which a comment is made after "Filename: [Added/Removed/Deleted]."
(CommitMessageParser.delete_trailing_blank_lines):
    - Fix class method arguments

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/commit.py:
(Commit.parser): Add --update/--no-update flags
(Commit.main): Set WKSCMPY_UPDATE_CHANGELOG env variable

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/setup.py:
(Setup.git): Set webkitscmpy.auto-update-changelog.

* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/commit_parser_unittest.py: Added.
(TestCommitParser):
(TestCommitParser.test_basic_commit):
* Tools/Scripts/libraries/webkitscmpy/webkitscmpy/test/setup_unittest.py:

Canonical link: https://commits.webkit.org/295625@main
